File: frontend/src/lib/sockets/ws-protocol.ts
/**
 * Chat WebSocket wire protocol.
 *
 * Every frame is { type: WsEvent, payload: ... }. The event enum is
 * mirrored in backend/server/core/ws_protocol.py — keep the two in sync.
 */

export enum WsEvent {
  Content = 'content',
  Metadata = 'metadata',
  Context = 'context',
  Error = 'error'
}

export interface WsMetadataPayload {
  totalTokens?: number
  promptTokens?: number
  completionTokens?: number
  totalCost?: number
  [key: string]: unknown
}

/** Retrieved chapter as sent by the backend (snake_case wire format). */
export interface WsContextChapter {
  chapter_id: string
  chapter_tag: string
  chapter: string
  subchapter: string
  title?: string
  rerank_score: number
}

export interface WsErrorPayload {
  detail: string
  error_id: string
  step: string
  code: number
}

export type WsFrame =
  | { type: WsEvent.Content; payload: string }
  | { type: WsEvent.Metadata; payload: WsMetadataPayload }
  | { type: WsEvent.Context; payload: WsContextChapter[] }
  | { type: WsEvent.Error; payload: WsErrorPayload }

export function parseWsFrame(raw: string): WsFrame | null {
  try {
    const data = JSON.parse(raw)

    if (
      data &&
      typeof data === 'object' &&
      Object.values(WsEvent).includes(data.type)
    ) {
      return data as WsFrame
    }

    return null
  } catch {
    return null
  }
}

File: backend/server/core/ws_protocol.py
"""Chat WebSocket wire protocol.

Every frame is {"type": WsEvent, "payload": ...}. The event enum is
mirrored in frontend/src/lib/sockets/ws-protocol.ts — keep the two in sync.
"""
import json
from enum import Enum
from typing import Any


class WsEvent(str, Enum):
    CONTENT = "content"      # payload: str — one streamed answer token/chunk
    METADATA = "metadata"    # payload: dict — token usage / cost for the turn
    CONTEXT = "context"      # payload: list[dict] — retrieved chapters for the answer
    ERROR = "error"          # payload: dict — detail, error_id, step, code


def ws_frame(event: WsEvent, payload: Any) -> str:
    return json.dumps({"type": event.value, "payload": payload})
